Don't do unnecessary refine in reinit

lindsayad opened the pull request:
Closes #30725

lindsayad committed:
Don't refine in reinit

Closes #30725

lindsayad committed:
Try to minimize the number of element loops that we do

But of course this broke a test of Exodus output involving syncing
between reference and displaced meshes. Rage

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2946545/), step Docs: sync website on 11e612f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30742/site)

This comment will be updated on new commits.

lindsayad committed:
Can just use libMesh System for DisplacedSystem

This saves 3 vectors per displaced system

lindsayad committed:
Cleanup meshChanged interface

lindsayad committed:
Pick next node/elem ID using max_node/elem_id()

If renumbering is disabled then using _mesh->n_nodes() to pick the ID
might actually land you in the middle of a `_nodes` vector since it
can contain null pointers

lindsayad committed:
Make Exodus handling of renumbering more robust

It shold be prepared to handle renumbering at any output event, not
just the initial output

lindsayad committed:
Allow XMFEM to still contract the mesh

lindsayad committed:
Add back no-arg meshChanged for compatibility with Griffin

moosebuild commented:
All [jobs](https://civet.inl.gov/event/239561/) on c73776c : invalidated by @lindsayad

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2962772/) on c73776c : invalidated by @lindsayad

GiudGiud commented on framework/src/outputs/Exodus.C:
```diff
@@ -533,19 +539,23 @@ Exodus::clear()
+        "The only way we shouldn't have contiguous numbering is if we've disabled renumbering");
+    other_mesh.allow_renumbering(true);
+    other_mesh.renumber_nodes_and_elements();
+    // Copying over the comment in MeshOutput::write_equation_systems
+    // Not sure what good going back to false will do here, the
+    // renumbering horses have already left the barn...
+    other_mesh.allow_renumbering(false);
```
if I see a mesh with allow_renumbering = false I kind of expect it did not get renumbered

lindsayad replied:
This is what is done in the `MeshOutput::write` code. I think the overriding priority is to keep the two meshes in sync

lindsayad replied:
And the value of this boolean does in fact change the outcome of calls to `ReplicatedMesh::renumber_nodes_and_elements`

lindsayad replied:
You can make an issue/PR to change the libmesh behavior if you wish but it will likely lead to exodiffs

lindsayad replied:
The bottom line is that as it stands now you shouldn't see `allow_renumbering(false)` and trust that means the mesh has not or will not be renumbered. `contract` for instance, which is always called post-adaptivity, always calls `renumber_nodes_and_elements` which always renumbers regardless of `_skip_renumber_nodes_and_elements`

GiudGiud replied:
> I think the overriding priority is to keep the two meshes in sync

Agreed
Good to know on the other items. There are two discussions posts open complaining about renumbering btw

lindsayad replied:
Associated libmesh issues: https://github.com/libMesh/libmesh/issues/3966, https://github.com/libMesh/libmesh/issues/1216, and https://github.com/libMesh/libmesh/issues/938

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7920,7 +7925,22 @@ FEProblemBase::adaptMesh()
-      meshChangedHelper(true); // This may be an intermediate change
+      meshChanged(
+          /*intermediate_change=*/true, /*contract_mesh=*/true, /*clean_refinement_flags=*/true);
+      // Once vectors are restricted, we can delete
+      // children of coarsened elements
+      _mesh.getMesh().contract();
```
didnt we just contract above?

lindsayad replied:
```suggestion
```

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7920,7 +7925,22 @@ FEProblemBase::adaptMesh()
+      {
+        _displaced_mesh->getMesh().contract();
+        MeshRefinement displaced_refinement(_displaced_mesh->getMesh());
+        displaced_refinement.clean_refinement_flags();
```
and same right above?

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -8137,6 +8169,10 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
   _has_jacobian = false; // we have to recompute jacobian when mesh changed
+
+  // Now for backwards compatibility with user code that overrode the old no-arg meshChanged we must
+  // call it here
```
ok making it final would not really be backwards compatible

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1724,8 +1724,13 @@ class FEProblemBase : public SubProblem, public Restartable
    * Update data after a mesh change.
+   * Iff intermediate_change is true, only perform updates as
+   * necessary to prepare for another mesh change
+   * immediately-subsequent. If \p changed_throug_amr then we will
```
```suggestion
   * immediately-subsequent. 
```
It's not a parameter. Best describe the parameters. Then give advice on how to use it when in the context of AMR

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2403,6 +2408,11 @@ class FEProblemBase : public SubProblem, public Restartable
+  /**
+   * Deprecated. Users should switch to overriding the meshChanged which takes arguments
+   */
+  virtual void meshChanged() {}
```
nvmd

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7920,7 +7925,22 @@ FEProblemBase::adaptMesh()
+      // Once vectors are restricted, we can delete
+      // children of coarsened elements
+      _mesh.getMesh().contract();
+      // Finally clean refinement flags so that if someone tries to project vectors again without
+      // an intervening mesh refinement to clean flags they won't run into trouble
+      MeshRefinement refinement(_mesh.getMesh());
+      refinement.clean_refinement_flags();
+      if (_displaced_mesh)
+      {
+        _displaced_mesh->getMesh().contract();
+        MeshRefinement displaced_refinement(_displaced_mesh->getMesh());
+        displaced_refinement.clean_refinement_flags();
+      }
+
```
```suggestion
```

lindsayad committed:
Remove duplicate code in FEProblemBase::adaptMesh

I think I did this at one point before I factored the change into `meshChanged` itself

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1724,8 +1724,13 @@ class FEProblemBase : public SubProblem, public Restartable
    * Update data after a mesh change.
+   * Iff intermediate_change is true, only perform updates as
+   * necessary to prepare for another mesh change
+   * immediately-subsequent. If \p changed_throug_amr then we will
+   * contract the mesh and clean refinement flags
```
```suggestion
   * immediately-subsequent. Note that after changing the mesh using AMR, contact_mesh and clean_refinement_flags should be set to true.
```

lindsayad committed:
Add more doxygen for meshChanged

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
failure unrelated and hopefully going to be fixed in https://github.com/idaholab/moose/pull/30844

moosebuild commented:
Job [Recover sweep evens](https://civet.inl.gov/job/2981916/) on 3f9e943 : invalidated by @lindsayad

moosebuild commented:
Job [Distributed mesh and recover sweep evens](https://civet.inl.gov/job/2981920/) on 3f9e943 : invalidated by @lindsayad

moosebuild commented:
Job [Recover sweep evens](https://civet.inl.gov/job/2981916/) on 3f9e943 : invalidated by @lindsayad
